### PR TITLE
[OBSV-847] feat: replace localStorage with backend user service

### PR DESCRIPTION
**Pull request by cseas**
## Description

[Jira](https://razorpay.atlassian.net/browse/OBSV-847) | [Test Environment](https://hypertrace-test.concierge.stage.razorpay.in/explorer) | [Swagger][1]

- [x] Modified the following methods to use the backend [Hypertrace User Service](https://hus.concierge.stage.razorpay.in/swagger/index.html) instead of localStorage.
  - Fetching all saved queries for a user
  - Renaming a query
  - Deleting a query
  - Saving a query
- [x] Added an `OnInit` method to both the "Explorer" and "Saved Queries" pages that will
  - check for any existing saved queries in localStorage
  - make a POST call to save each query to the backend user service
  - delete the queries saved in localStorage
- [x] The `user/add` endpoint is called on every application load once, right after calling `/user-info`
- [x] The `UserPreferenceService` is used for interacting with the User Service backend APIs.
- [x] `UserModule` is replaced by `UserPreferenceModule`

### Testing
Related tests have been modified.
- [x] Tested on test environment with default user
- [ ] Tested on stage environment with real user

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

[1]: https://hus.concierge.stage.razorpay.in/swagger/index.html

### CODE REVIEW

File: projects/observability/src/pages/explorer/explorer.types.ts
Comment (cseas) [inline projects/observability/src/pages/explorer/explorer.types.ts:1]: > Why move this type to a separate file? 

✅  To fix the circular structure of imports.

Screenshot 2022-08-05 at 12 09 21 PM

Ideally any exported types should be in separate files as a rule of thumb. Not sure why the Hypertrace team keeps everything in one file, leads to a lot of cases where 2 files are importing something from each other.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [inline projects/observability/src/pages/saved-queries/saved-queries.service.ts:10]: @akashmane1598 Please give me a heads up when the `/user-preferences` endpoint starts working. Will change this.

File: src/app/shared/add-user/add-user.service.ts
Comment (cseas) [inline src/app/shared/add-user/add-user.service.ts]: This file is temporary. Since the `addUser()` method interacts with the `/user-preferences` endpoint, this method should ideally be added to the [UserPreferenceService][1] from #56. Will add the change when the service is available in `rzp_main`.

[1]: https://github.com/razorpay/hypertrace-ui/pull/56/files#diff-d0ef16a7e639d1c040567b66d3c9f391f6543c671838fe8689f6dfc43d8f8292R13

File: src/app/application-frame/application-frame.component.ts
Comment (cseas) [inline src/app/application-frame/application-frame.component.ts]: > Why call this method here and not in the `UserInfoService`?

As explained in another comment, the `addUser()` method will be added to the `UserPreferenceService`, which uses `UserInfoService` in its constructor.

To call `this.userPreferenceService.addUser()` inside `load()` of `UserInfoService`, we'll need to add `UserPreferenceService` inside the constructor of `UserInfoService`. 

Not sure if that'll work, having 2 services that both depend on each other; scenario explained in below code snippet.

```js
class UserInfoService {
  constructor(private readonly userPreferenceService: UserPreferenceService) {}
  
  load() {
    this.userPreferenceService.addUser();
  }
}

class UserPreferenceService {
  constructor(private readonly userInfoService: UserInfoService) {
    // user email is required by all methods in this service
  }
}
```
Also seems to violate [Single Responsibility Principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#single-responsibility-principle).

File: projects/observability/src/pages/explorer/explorer-service.test.ts
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer-service.test.ts:81]: @cseas Why are we removing this test case?

File: projects/observability/src/pages/explorer/explorer-service.ts
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer-service.ts:53]: Is service taking ownership of verifying duplication? Or we are not doing a duplication check now?

File: projects/observability/src/pages/explorer/explorer.component.ts
Comment (jaywalker21) [inline projects/observability/src/pages/explorer/explorer.component.ts]: Can we add a `// TODO` comment here please? So it ensures we delete it after the initial 4-5 days.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (jaywalker21) [inline projects/observability/src/pages/saved-queries/saved-queries.service.ts]: Why do we need a `static` variable here? Can't we read it in a `const` scoped to this class?

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (jaywalker21) [inline projects/observability/src/pages/saved-queries/saved-queries.service.ts]: Can we inject/read this the same we way we are doing this for `/graphql`

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (jaywalker21) [inline projects/observability/src/pages/saved-queries/saved-queries.service.ts]: What do you folks think about adding the default header via [interceptor](https://angular.io/guide/http)?

I think it might look clean and easy on the eye.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (jaywalker21) [inline projects/observability/src/pages/saved-queries/saved-queries.service.ts:52]: Just checking that you played around with different scenarios, `undefined` value especially for this.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (jaywalker21) [inline projects/observability/src/pages/saved-queries/saved-queries.service.ts]: Why did we go with this weird naming style?
Shouldn't Shivam be sending the first char in `smallcase` and everything be like `snakeCase`?

File: src/app/application-frame/application-frame.component.ts
Comment (jaywalker21) [reply src/app/application-frame/application-frame.component.ts]: I think calling it that way is required. Can't we have a method in `UserPreferenceService` which is called in `APP_INITIALIZE`, just the way we call `load()` for `UserInfoService`. That method will load the data and store the preferences in the future within the `UserPreferenceService` itself.

I don't think adding in `application-frame.component` will help in the future.

// @cseas @akashmane1598

**Review by jaywalker21 — COMMENTED** [review]
Added few comments

File: projects/observability/src/pages/explorer/explorer-service.test.ts
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer-service.test.ts:81]: ℹ️  On the frontend, we can only test whether the APIs are being called with the correct parameters. Testing actual functionality of the APIs is beyond the scope of frontend. The duplication logic was earlier present in the frontend. Now it's upto the backend to check duplicates if they want to.

File: projects/observability/src/pages/explorer/explorer-service.ts
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer-service.ts:53]: ℹ️  The backend API doesn't do duplication checks. It's possible to save the same query twice. Never understood the use case but it seems the backend team is of the opinion "if someone wants to save the same query again with a different name, then let them".

File: projects/observability/src/pages/explorer/explorer.component.ts
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: Many users don't use Hypertrace everyday. Keeping this around for a couple months will ensure everyone gets a smooth transition without noticing. Have added a comment explaining everything in the method definition. Will add a comment here too.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: ❓  Any instances of this class would need the same email value for a user session. It's a value we fetch once and then treat it like a `const` so making it `static` seemed appropriate. Should I remove the `static`?

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: Yes, but I was thinking that'd be throwaway work since the methods here will be replaced by UserPreferenceService, that handles the token injection.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts:52]: ℹ️  The second argument of `preferenceService.get()` specifies the default return value so there shouldn't be an undefined case here. Even if all localStorage is wiped, the subscription will still return an empty array.

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: Not sure about backend naming conventions of our team. Just added these fields as they're coming in.
https://hus.concierge.stage.razorpay.in/swagger/index.html

Screenshot 2022-08-10 at 1 33 06 PM

File: src/app/application-frame/application-frame.component.ts
Comment (cseas) [reply src/app/application-frame/application-frame.component.ts]: ❓ Not sure how `APP_INITIALIZE` handles ordering of calls. If we do that, then we need to make sure `APP_INITIALIZE` calls `UserInfoService` before `UserPreferenceService`, otherwise if it happens in the reverse order, then `UserInfoService` will return the default `ht-user`.

Maybe we can remove `APP_INITIALIZE` from `UserInfoService` altogether and move it to `UserPreferenceService`?

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: `UserPreferenceService` from #56 handles the email header, waiting for it to be merged to `rzp_main`. Or should I copy that file to my PR to make the change now?

File: projects/observability/src/pages/explorer/explorer.component.ts
Comment (cseas) [reply projects/observability/src/pages/explorer/explorer.component.ts]: ✅ Added todo comment

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: Waiting for Shivam to change the fields to camelCase:
https://razorpay.slack.com/archives/CU5GKS8MQ/p1660120125275039?thread_ts=1660119417.491239&cid=CU5GKS8MQ

Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: ✅ Changed to camelCase

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: ✅ Replaced all http calls with UserPreferenceService

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: ✅ Replaced all http calls in this file with `UserPreferenceService`

File: src/app/application-frame/application-frame.component.ts
Comment (cseas) [reply src/app/application-frame/application-frame.component.ts]: @jaywalker21 As assumed above, circular dependencies are not allowed in services. I tried adding `this.userPreferenceService.get('/v1/user/add');` to the `load()` method of `UserInfoService` and got this error.

Screenshot 2022-08-10 at 7 31 24 PM

File: src/app/shared/add-user/add-user.service.ts
Comment (cseas) [reply src/app/shared/add-user/add-user.service.ts]: ✅ File deleted

File: projects/observability/src/pages/saved-queries/saved-queries.service.ts
Comment (cseas) [reply projects/observability/src/pages/saved-queries/saved-queries.service.ts]: ✅ Removed `userEmail` from this class entirely. It's not needed anymore.

File: src/app/application-frame/application-frame.component.ts
Comment (cseas) [reply src/app/application-frame/application-frame.component.ts]: ✅ Removed `APP_INITIALIZER` from `UserInfoService` and added it to `UserPreferenceService` instead which calls both `/user-info` and `/user/add` now.

**Review by jaywalker21 — APPROVED** [review]
Looks good to me.
@akashmane1598 Will you also have a look once before this is merged?

**Review by akashmane1598 — APPROVED** [review]
LGTM

**Pull request merged**
